### ISSUE TITLE
plot with no content

### ISSUE THREAD

**Issue by XYZ-NCI** [broadinstitute/infercnv#231]
Hi! Thank you so much for this great tool. I have been using it since last summer and found it really helpful. Thanks!

The tool has always been great and plots generated well until recently, that all of the png plots turned out to be empty. For my most recent runs (R/3.6.0 and infercnv/1.3.2), all of the png files are empty. It looks to me that the other files (obj, txt, dat) were all generated well. Do you have any idea of the potential reasons? I'm happy to provide more information if needed.

Thanks again!
XY

**Comment by GeorgescuC** [comment]
Hi @XYZ-NCI ,

Are you using infercnv on your local machine or a server? The issue may be coming from the graphical libraries used to generate the images.
If you set `output_format="pdf"`, do the pdfs look fine?

Regards,
Christophe.

**Comment by XYZ-NCI** [comment]
Hi Christophe,

Thank you so much for your immediate attention. I am using infercnv on a server. Since my data are so big, it’s typically memory consuming and time consuming to run one big dataset. I have not tried pdf yet (I saw previous comments about the size of pdf). I will give it a try and let you know.

Thanks a lot!
Xiyuan

**Comment by GeorgescuC** [comment]
Hi @XYZ-NCI ,

Then the issue may be related to an update of the R/graphical dependencies on the server you are using. R has for example released their new version 4.0 3 days ago, and some dependencies may have been updated even though you are still using R 3.6.0. If your server allows it, you could also give a try to [our docker/singularity images](https://github.com/broadinstitute/inferCNV/wiki/Installing-infercnv#using-a-container-image).

Regards,
Christophe.

**Comment by XYZ-NCI** [comment]
Hi Christophe,

Thanks for the recommendation.
I have downloaded pre-built image from https://data.broadinstitute.org/Trinity/CTAT_SINGULARITY/InferCNV/ into our HPC and tried to run my script:
#!/bin/sh
# This file runs Singularity
set -e
module load singularity
singularity exec -e -B `pwd` path_to_it/infercnv.1.3.5.simg Rscript path_to_the_r_script/inferCNV_test.r

And the r script is as following:
library(Seurat)
library(infercnv)
data = readRDS("path_to_data/data.rds")

datamtx_data = as.matrix(GetAssayData(data, slot = "counts"))
info_data = rownames(data@meta.data)
tumor_info_data = gsub("_.*$", "", info_data)
tumor_info_data = as.data.frame(tumor_info_data, row.names = info_data)
infercnv_obj_data = CreateInfercnvObject(
  raw_counts_matrix = datamtx_data,
  annotations_file = tumor_info_data,
  delim = "\t",
  gene_order_file = "gencode_v19_gene_pos.txt",
  ref_group_names = c("X", "Y", "Z")
)
infercnv_obj_data = infercnv::run(infercnv_obj_data,
                                                             cutoff=0.1,
                                                             out_dir="cluster_by_group_data",
                                                             cluster_by_groups=TRUE,
                                                             denoise=TRUE,
                         HMM=TRUE)

It couldn’t run… I’m very new to Singularity. Could you kindly point to me how would I run this r script?

Thanks a lot!
Xiyuan

**Comment by GeorgescuC** [comment]
Hi @XYZ-NCI ,

Does running the command generate any output?

The issues I see that could be happening are:
- The Rscript or the files you load (`path_to_data/data.rds`, `gencode_v19_gene_pos.txt` and `path_to_the_r_script/inferCNV_test.r`) are not under `pwd`, so they do not get linked inside the singularity image, which prevents R from accessing them. 
- There is a conflict between the libraries you have installed locally and the ones inside the image because singularity sources your home directory by default. You can add the `-H /path/to/some/empty/folder` to make it use an empty folder as the home from outside the image.

Regards,
Christophe.

**Comment by XYZ-NCI** [comment]
Hi Christophe,

Thank you so much for your timely reply. No, there was no output… And the error message is:
[+] Loading singularity  3.5.3  on cn1647
Bioconductor version 3.11 (BiocManager 1.30.10), ?BiocManager::install for help
Error in library(Seurat) : there is no package called ‘Seurat’
Execution halted

I don’t think I have successfully started the execution at all. I’m sorry but Singularity is a brand new concept for me…
Best,
Xiyuan

**Comment by GeorgescuC** [comment]
Hi @XYZ-NCI ,

It seems like there is an issue with the Seurat installation in that image. I generated a new image at https://data.broadinstitute.org/Trinity/CTAT_SINGULARITY/InferCNV/infercnv.1.3.6-S.simg where Seurat should work properly if you can try using that one instead.

Regards,
Christophe.

**Comment by XYZ-NCI** [comment]
Thank you so much for checking! I will give it a try and let you know.

Regards,
Xiyuan

**Comment by XYZ-NCI** [comment]
Just gave it a try. Now loading Seurat seems working but my run still failed. Got the error message:
Error in gzfile(file, "rb") : cannot open the connection
Calls: readRDS -> gzfile
In addition: Warning message:
In gzfile(file, "rb") : cannot open compressed file ' ', probable reason 'No such file or directory'

Is it possible that within the container, there is no connection to other directories, where I stored the RDS files with the expression data from Seurat?

Thanks again!
Xiyuan

**Comment by GeorgescuC** [comment]
Hi @XYZ-NCI ,

If the folder is not under your current working directory (pwd), you need to add the path to the -B argument in the singularity command you run. For example, you can set it to ```-B `pwd`,/path/to/rds/```.

Regards,
Christophe.

**Comment by XYZ-NCI** [comment]
Hi Christophe,

I’m trying this now. Will let you know when I get some result.

Another unrelated questions: is there any way to perform some phylogeny analysis using the generated result from inferCNV? I tried the Tumor subclustering methods<https://github.com/broadinstitute/infercnv/wiki/infercnv-tumor-subclusters> as you recommended in the Wiki but am wondering if you can further perform/visualize the result in a more intuitive way? More specifically, can we feed some intermediate file or the final.object generated in inferCNV to other packages to plot a phylogenic tree and still know what’s in the tree?

Thanks a lot!
Xiyuan

**Comment by GeorgescuC** [comment]
Hi @XYZ-NCI ,

When a figure is generated, a newick tree file is generated with the dendrogram displayed on the side of the figure where cell names are indicated. There is one line per group in case of multiple groups (top level branches).

You can also pull that information as hclust objects from the `infercnv_obj@tumor_subclusters$hc` field of the infercnv object.

Regards,
Christophe.

**Comment by XYZ-NCI** [comment]
Hi Christophe,

I’m actually trying to find the subclones within the tumor sample and am doing so with the “subcluster mode”. It seems that it will take a while to run each sample but I just want to confirm with you that using the following setting, I will get the “tree branching” information (with cells assignment) in “HMM_CNV_predictions.HMMi6.rand_trees.hmm_mode-subclusters.Pnorm_0.5.cell_groupings” file.
infercnv_obj = infercnv::run(infercnv_obj,
                                                             cutoff=0.1,
                                                             out_dir="cluster_by_group",
                                                             cluster_by_groups=TRUE,
                                                             denoise=TRUE,
                                                             analysis_mode='subclusters',
                                                             hclust_method='ward.D2',
                                                             tumor_subcluster_partition_method='random_trees',
                                                             tumor_subcluster_pval=0.05,
                         HMM=TRUE)
Basically, I would like to get information of “how the tree branches” (1, 1.1, 1.1.1 etc) to be used in UPhyloplot2 to generate an “evolution tree”. When I set tumor_subcluster_partition_method=’qnorm', I can find grouping information in .cell_groupings file but it’s only one level of subgroups (s1, s2, s3, etc). I was wondering, by setting “tumor_subcluster_partition_method='random_trees',” do I get the grouping info like (1, 1.1, 1.1.1 etc)? or do I need to tweak anything else? Thanks a lot!

Btw, I found out the “png plot with no content” problem is probably due to the size of the data. I used the same setting with the infercnv that was installed on our serve to run a much smaller dataset, all plots were generated fine.

Regards,
Xiyuan

**Comment by GeorgescuC** [comment]
Hi @XYZ-NCI ,

To get the tree itself, you can access the `infercnv_obj@tumor_subclusters$hc` field and each annotation group will have an hclust object you can use to extract the data in the format you want. For which cells are part of which split in a tree, you can find their names in `infercnv_obj@tumor_subclusters$subclusters` where each annotation group will have fields with the branches subdivisions appended at the end pointing the the list of cells.

For the plots, it may be because of RAM issues then, because even for big matrices, although it can take a while, it should produce proper figures if the process is not interrupted. You can also try updating infercnv to a newer version (1.3.5 or 1.4.0) as version 1.3.3 introduced large speed improvements to the plotting for large matrices (with a few fixes in the following versions).

Regards,
Christophe.
